File: Top Interview 150/Java Code/matrix/54-spiral-matrix.java
// Time:O(m*n)
// space:O(1)
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer>list=new ArrayList<>();
        int m=matrix.length;
        int n=matrix[0].length;
        int minc=0,minr=0,maxr=m-1,maxc=n-1;
        int cnt=0;int total=m*n;
        while(cnt<total){
            // top
            for(int i=minr,j=minc;j<=maxc && cnt<total;j++){
                list.add(matrix[i][j]);
                cnt++;
            }
            minr++;
            //right wall
            for(int i=minr,j=maxc;i<=maxr && cnt<total;i++){
                 list.add(matrix[i][j]);
                cnt++;
            }
            maxc--;
            // bottom
             for(int i=maxr,j=maxc;j>=minc  && cnt<total;j--){
                 list.add(matrix[i][j]);
                cnt++;
            }
            maxr--;
            // left
            for(int i=maxr,j=minc;i>=minr  && cnt<total;i--){
                list.add(matrix[i][j]);
                cnt++;
            }
            minc++;
            
        }
        return list;

    }
}

File: Top Interview 150/C++ Code/Matrix/54-spiral-matrix.cpp
// Time:O(m*n)
// space:O(1)
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& a) {
        vector<int>v;
         int n=a.size();
        int m=a[0].size();
        int minc=0,minr=0,maxr=n-1,maxc=m-1;
        int count=0;
        int total=m*n;
        while(count<total){
            // top
             for(int i=minr,j=minc ;j<=maxc && count<total;j++){
                 v.push_back(a[i][j]);
                 count++;
             }
            minr++;
            // right
            for(int i=minr,j=maxc ;i<=maxr && count<total;i++){
                 v.push_back(a[i][j]);
                 count++;
             }
            maxc--;
            // bottom side
            for(int i=maxr,j=maxc ;j>=minc && count<total;j--){
                 v.push_back(a[i][j]);
                 count++;
             }
            maxr--;
            // left side
            for(int i=maxr,j=minc;i>=minr && count<total;i--){
                 v.push_back(a[i][j]);
                 count++;
             }
            minc++;
            
        }
        return v;
    }
};
